### ISSUE TITLE
[question] "inheriting" dynamic options of dependencies (requires)?

### ISSUE THREAD

**Issue by iiknd** [conan-io/conan#8693]
Hi,

we have the following scenario:

ModuleA:
- a **lot** of options how this can be configured
- options may vary between different versions of ModuleA
- we want to dynamically set the "options = { ... }" for the conanfile.py, parsed from a file inside the ModuleA's .git repository

ModuleB
- we want to have the **same** options declared here as for ModuleA

ModuleC, ModuleD, ModuleE ....
- repeats regarding the options

A <- B <- C <- D .....

The problem is that the options dictionary is populated during "conan export" and the other modules besides ModuleA do not have direct access to the moduleA.git:some_file.json where the options can be parsed from.

Is there any way for a conanfile.py to "query" the options dictionary of a Conan package it "requires"?
If this is possible is it possible to use the RREV of the ModuleA in conjunction?

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Hi @unzap 

I am afraid it is physically impossible. If you have a conanfile.txt that lets say, requires ``pkgc/1.0``. First step is to locate it, fetch/download it, and then load the conanfile.py inside the package. Then Conan needs to know the requirements of ``pkgc/1.0``. As conditional requirements like:
```python
def requirements(self):
     if self.options.with_ssl:
         self.requires("openssl/1.2")
```
Are possible (not only possible, but wildly popular), then the options of ``pkgc/1.0`` are needed at this point. 
Now note that its dependencies like ``pkgb/1.0`` are completely unknown. They might not even be in the filesystem of this machine. It simply doesn't exist yet. Not only the options, the revision, etc., it is impossible to know anything about a dependency that is not a dependency yet.

So, no, it is not possible.
If you are using the same set of options in many Conan packages, this might totally be the case of custom ``settings``. Have you considered that approach? What would be your use case for those options?

**Comment by iiknd** [comment]
> If you are using the same set of options in many Conan packages, this might totally be the case of custom settings. Have you considered that approach?

Wouldn't this prevent uploading the packages in conan-center? To my understanding all packages in conan-center must stick to the upstream version of settings.yml?

> What would be your use case for those options?

E.g. feature "enable tracing" should be "inherited" trough the whole requires chain if defined by ModuleA.
There are many similar options. And we'd like to have one source of truth, i.e. one
file that contains the possible options to avoid duplicating them in multiple places.

And maybe the most important reason being that once the options are updated in ModuleA it would mean updating the conanfile.py in all of the recipes trough the requires chain. We have recipes in the same .git repository with the source code, separate .git repositories per module. So updating the options in ModuleA triggers a lot of changes in other .git modules.

**Comment by memsharded** [comment]
> Wouldn't this prevent uploading the packages in conan-center? To my understanding all packages in conan-center must stick to the upstream version of settings.yml?

Yes, ConanCenter cannot use custom settings. I assumed it was more a proprietary scenario.

> E.g. feature "enable tracing" should be "inherited" trough the whole requires chain if defined by ModuleA.
There are many similar options. And we'd like to have one source of truth, i.e. one
file that contains the possible options to avoid duplicating them in multiple places.

Options cannot be propagated that way, because they are specifically designed to have potentially different values for different packages. And because of the above explanation, that they cannot propagate down, because the upstream doesn't exist yet.

Still it is possible to consistently define them with the same value, for all packages that define it. Define in your consumer:

```
[options]
*:enable_tracing=True
```

Similar syntax can be applied in conanfile.py, in profiles, as well. This will define the same value for all dependencies, easily, without needing to change recipes at all.

**Comment by memsharded** [comment]
By the way, if you could please point to examples in https://github.com/conan-io/conan-center-index repo, that would be useful to understand the issue, I cannot find the ``enable_tracing`` use case there.

**Comment by AbrilRBS** [comment]
Closing as stale as we didn't hear back from you in a while. If anyone faces similar issues, please feel free to create a new issue
